thorlcr/master/thmastermain.cpp

```diff
@@ -338,27 +338,86 @@ class CRegistryServer : public CSimpleInterface
     {
         CriticalBlock block(crit);
         unsigned i=0;
-        for (; i<queryNodeClusterWidth(); i++)
+        unsigned numNodes = queryNodeClusterWidth();
+        unsigned numReplyNodes = 0;
+        mptag_t shutdownTag = createReplyTag();
+        for (; i<numNodes; i++)
         {
             if (status->test(i))
             {
-                status->set(i, false);
                 SocketEndpoint ep = queryNodeGroup().queryNode(i+1).endpoint();
-                if (watchdog)
-                    watchdog->removeSlave(ep);
                 CMessageBuffer msg;
                 msg.append((unsigned)Shutdown);
+                msg.append((unsigned)shutdownTag);
+#ifdef _DEBUG
+                StringBuffer tmp;
+                ep.getUrlStr(tmp);
+                DBGLOG("sending Shutdown:%u msg to slave %d (%s)", (unsigned)shutdownTag, i+1, tmp.str());
+#endif
                 try
                 {
-                    queryNodeComm().send(msg, i+1, masterSlaveMpTag, MP_ASYNC_SEND);
+                    if (queryNodeComm().send(msg, i+1, masterSlaveMpTag, MP_ASYNC_SEND))
```

Comment:
Are you saying do not use numNodes = queryNodeClusterWidth() as the original loop iter code does, but use slavesRegistered instead ?